### PR TITLE
Hotfix/add abseil validation

### PR DESCRIPTION
**Pull request by klimkin**
Specify library name and version:  **abseil/20210324.2**

* Add package validation to match supported platforms from https://abseil.io/docs/cpp/platforms/platforms
* Fix test_package to match C++ standard with upstream

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/abseil/all/conanfile.py [hunk 1]
```diff
@@ -35,9 +35,19 @@ def config_options(self):
...
+        version = tools.Version(self.settings.compiler.version)
+        compiler = self.settings.compiler
+        if compiler == "gcc" and version < "5.1":
+            raise ConanInvalidConfiguration("abseil requires gcc 5.1+")
+        if compiler == "clang" and version < "3.7":
+            raise ConanInvalidConfiguration("abseil requires clang 3.7+")
+        if compiler == "apple-clang" and version < "7.3":
+            raise ConanInvalidConfiguration("abseil requires Xcode 7.3.1+")
+        if compiler == "Visual Studio" and version < "14":
+            raise ConanInvalidConfiguration("abseil requires at least Visual Studio 2015 Update 3")
```
Comment (SpaceIm) [inline recipes/abseil/all/conanfile.py:50]: why being so strict. We know it compiles with gcc 4.9 and clang 3.4

File: recipes/abseil/all/test_package/conanfile.py [hunk 2]
```diff
@@ -7,6 +7,8 @@ class TestPackageConan(ConanFile):
 
     def build(self):
         cmake = CMake(self)
+        if not self.settings.compiler.cppstd:
+            cmake.definitions["CMAKE_CXX_STANDARD"] = 11
```
Comment (SpaceIm) [inline recipes/abseil/all/test_package/conanfile.py:11]: You can use target_compile_features in CMake instead.

Comment (SpaceIm) [inline recipes/abseil/all/test_package/conanfile.py:11]:
```suggestion
```

File: recipes/abseil/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -8,9 +8,7 @@ conan_basic_setup(TARGETS)
 find_package(absl REQUIRED)
 add_executable(${PROJECT_NAME} test_package.cpp)
 target_link_libraries(${PROJECT_NAME} absl::strings absl::flat_hash_map absl::flat_hash_set absl::int128 absl::time)
```
Comment (SpaceIm) [inline recipes/abseil/all/test_package/CMakeLists.txt:10]:
```suggestion
target_link_libraries(${PROJECT_NAME} absl::strings absl::flat_hash_map absl::flat_hash_set absl::int128 absl::time)
target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_11)
```

File: recipes/abseil/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -8,9 +8,7 @@ conan_basic_setup(TARGETS)
...
-set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 11)
 
 # Test also global target
 add_executable(${PROJECT_NAME}_global test_package.cpp)
 target_link_libraries(${PROJECT_NAME}_global CONAN_PKG::abseil)
```
Comment (SpaceIm) [inline recipes/abseil/all/test_package/CMakeLists.txt:14]:
```suggestion
target_link_libraries(${PROJECT_NAME}_global CONAN_PKG::abseil)
target_compile_features(${PROJECT_NAME}_global PRIVATE cxx_std_11)
```

**Review by SpaceIm — COMMENTED** [review]
I would prefer target_compile_features with cxx_std_11 (CMake >= 3.8)

File: recipes/abseil/all/test_package/conanfile.py (hunk 2 above)
Comment (klimkin) [reply recipes/abseil/all/test_package/conanfile.py:11]: Yes, it's broken if compiling with standard c++17.

File: recipes/abseil/all/conanfile.py (hunk 1 above)
Comment (klimkin) [reply recipes/abseil/all/conanfile.py:51]: Here is example user code builds but doesn't links with gcc 4.9. Maybe you can see if there is a problem with the user code, I don't have gcc 4.9 to check.
https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/6330/7/bloaty/1.1//summary.json


BUILT + TEST_OK | bloaty/1.1 | c9aa50b64a4a418b96f07ff2342874cc18e9dc67 | Linux | Linux | x86_64 | x86_64 | gcc | libstdc++ | 4.9 |   | Release | profile | build | test
-- | - | - | - | - | - | - | - | - | - | - | - | - | - | -
FAILED | bloaty/1.1 | 5e02e2869869a374987d358efbe9f1126c0a502b | Linux | Linux | x86_64 | x86_64 | gcc | libstdc++ | 4.9 |   | Debug | profile | build

**Comment by Wuestenschiff** [comment]
Can we speed this up. As actually broken for all projects not using c++11 gives thows `std::bad_alloc` right and left

**Comment by SpaceIm** [comment]
@Wuestenschiff Not sure to understand, this PR doesn't change anything for consumers, it fixes test package, that's all.

**Comment by Wuestenschiff** [comment]
> @Wuestenschiff Not sure to understand, this PR doesn't change anything for consumers, it fixes test package, that's all.

Just realised that... Still cppstd needs to be fixed

**Comment by SpaceIm** [comment]
@Wuestenschiff In abseil recipe, if you replace

```python
if not self.settings.compiler.cppstd:
    self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
```

by

```python
if not tools.valid_min_cppstd(self, 11):
    self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
```

is it better?

**Comment by Wuestenschiff** [comment]
> @Wuestenschiff In abseil recipe, if you replace
> 
> ```python
> if not self.settings.compiler.cppstd:
>     self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
> ```
> 
> by
> 
> ```python
> if not tools.valid_min_cppstd(self, 11):
>     self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
> ```
> 
> is it better?


I am not sure how to achieve this (sorry I am new to conan - found only documentation to copy recipes 1:1 not to change them) also tried to set cppstd in the profile no luck so far. (Also sorry for hijacking this bug... was not my intention but as I have already lost a lot of time I really apprichiate your insights). If you are interested my setup is really simple gcc 11 or clang 12 `CMAKE_CXX_STANDARD 20` just compiling the example code listed here: https://github.com/googleapis/google-cloud-cpp#quickstart This will thow `std::bad_alloc` with all cppstd except 11

**Comment by klimkin** [comment]
I think I finally understand the problem of forcing C++ standard. If headers use `#ifdef` to decide on the types in API, mismatch in C++ standard between any upstream/downstream will result in link issues. The best solution is not to force the standard, but to let Conan manage the value for all packages.

File: recipes/abseil/all/conanfile.py [hunk 5]
```diff
@@ -47,8 +58,6 @@ def _configure_cmake(self):
         if self._cmake:
             return self._cmake
         self._cmake = CMake(self)
-        if not self.settings.compiler.cppstd:
-            self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
```
Comment (klimkin) [inline recipes/abseil/all/conanfile.py:51]: Breaks linkage, if settings don't set cppstd and downstream forces C++17.

File: recipes/abseil/all/test_package/CMakeLists.txt [hunk 6]
```diff
@@ -8,9 +8,7 @@ conan_basic_setup(TARGETS)
...
-set_property(TARGET ${PROJECT_NAME}_global PROPERTY CXX_STANDARD 11)
```
Comment (klimkin) [inline recipes/abseil/all/test_package/CMakeLists.txt:16]: Breaks linkage, if settings set cppstd to C++17, forcing Abseil to build with C++17.

File: recipes/abseil/all/conanfile.py (hunk 1 above)
Comment (klimkin) [reply recipes/abseil/all/conanfile.py:51]: @SpaceIm I think I have figured out the main issue related to the CPP_STANDARD mismatch between upstream and downstream. So, the validation is irrelevant. Please let me know if you want me to remove the adding checks.

File: recipes/abseil/all/conanfile.py (hunk 5 above)
Comment (klimkin) [reply recipes/abseil/all/conanfile.py:51]: @SpaceIm Please confirm if this makes sense.

File: recipes/abseil/all/test_package/CMakeLists.txt (hunk 6 above)
Comment (klimkin) [reply recipes/abseil/all/test_package/CMakeLists.txt:16]: @SpaceIm Please confirm if this makes sense.

File: recipes/abseil/all/conanfile.py [hunk 7]
```diff
@@ -47,8 +58,6 @@ def _configure_cmake(self):
         if self._cmake:
             return self._cmake
         self._cmake = CMake(self)
```
Comment (SpaceIm) [inline recipes/abseil/all/conanfile.py:60]:
```suggestion
        self._cmake = CMake(self)
        if not tools.valid_min_cppstd(self, 11):
            self._cmake.definitions["CMAKE_CXX_STANDARD"] = 11
```

File: recipes/abseil/all/test_package/CMakeLists.txt (hunk 3 above)
Comment (SpaceIm) [inline recipes/abseil/all/test_package/CMakeLists.txt:10]:
```suggestion
target_link_libraries(${PROJECT_NAME} absl::strings absl::flat_hash_map absl::flat_hash_set absl::int128 absl::time)
target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_11)
```

File: recipes/abseil/all/test_package/CMakeLists.txt (hunk 4 above)
Comment (SpaceIm) [inline recipes/abseil/all/test_package/CMakeLists.txt:14]:
```suggestion
target_link_libraries(${PROJECT_NAME}_global CONAN_PKG::abseil)
target_compile_features(${PROJECT_NAME}_global PRIVATE cxx_std_11)
```

**Review by SpaceIm — COMMENTED** [review]
Should fix your issues, and allow all configurations in CCI. The big problem with your approach is that you prevent CI to test configurations (and therefore provide pre-built packages) for which compiler default standard is C++98, and it's a big issue for apple-clang.

File: recipes/abseil/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/abseil/all/conanfile.py:51]: now you can remove these checks I think

**Review by SpaceIm — APPROVED** [review]

**Comment by SpaceIm** [comment]
As I said, CMake min version of test_package should have been bumped to 3.8, but no need to trigger CI again just fot that I think.

**Comment by SSE4** [comment]
> As I said, CMake min version of test_package should have been bumped to 3.8, but no need to trigger CI again just fot that I think.

I don't think we should set unnecessarily high CMake version requirement. if it works for 2.8, now reason to specify higher.

**Review by SSE4 — APPROVED** [review]

**Comment by SpaceIm** [comment]
> > As I said, CMake min version of test_package should have been bumped to 3.8, but no need to trigger CI again just fot that I think.
> 
> I don't think we should set unnecessarily high CMake version requirement. if it works for 2.8, now reason to specify higher.

`cxx_std_11` compile feature has been introduced in CMake 3.8

**Comment by uilianries** [comment]
Conan 2.0 approved CMake 3.15 as minimal version. Still, I think `set_property(... CXX_STANDARD 11)` is enough for test_package.

**Review by uilianries — APPROVED** [review]

**Comment by SpaceIm** [comment]
removing CXX_STANDARD is the purpose of this PR, since it seems to break when consumers set compiler.cppstd with another standard (abseil compiled with C++17 standard and test package with C++11).

**Pull request merged**